README.md
# DrySpy

This package can be used to detect open water from multi band optical satellite images and in a second step determine if a river runs dry. The water detection part can be used without the second part. It been tested with Sentinel-2, PlanetScope and PlanetFusion data.

1. The main class that does the open water detection of a single image is `co_water_detect.canny_otsu_water_detect`. The class is very generic and can be used for single images or with priori information in time series processing. It has been tested with Sentinel-2, PlanetScope and Planet Fusion data. The algorithm is based on the work of Donchyts et.al. 2016.
2. The `runner.py` file is a wrapper around the water detection class to perform river monitoring and detection of dry periods. If you install the `DrySpy` package (using pip) this wrapper will be installed as the `DrySpy` command-line program. This is a very specific runner, it is suggested _to write you own for most applications_.
3. The `rw_utils.py` file contains simple (draft) readers for Planet Fusion and PlanetScope and Sentinel-2 data and are used in the `runner.py` command line wrapper. These assume the data originate from a specific source (e.g. Sentinel-hub in case of the S2 data). 


## Installation

Starting from the dryspy folder:

- Setup the environment using conda: `conda env create -f env-dryspy.yml`
- activate the environment: ` conda activate dryspy`
- install the package: `pip install .`

If you want to use the jupyter notebook examples you will need to manually install the jupyter package (`conda install jupyter`).

## Starting

A simple example of using the `co_water_detect.canny_otsu_water_detect` class can be found in `notebooks/example/example-waterdetect.ipynb`. This example loads the class and some sentinel-2 data and performs water detection on nir and swir bands with different settings. 


## DrySpy

The `DrySpy` command can be used to loop over a stack (in time) of Planet Fusion images, detect water for each day and summarize the results. In addition, it tries to detect dry periods. It is a very specific piece of software that needs to be adjusted for other use cases. It may be easier to write your own wrapper/runner. 

```
 Usage: DrySpy [OPTIONS]

 Run the DrySpy program. Detect water in regions and calculate statistics

╭─ Options ──────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ --input_folder                 TEXT  input folder with planetscope data (PS, UDM and XML), FUSION (PF)  or     │
|                                      Sentinel-2 (S2) data                                                      │
│ --prior_water_occurrence       TEXT  input file (1 band tiff (0-1)) of prior water occurence                   │
│ --clip_file                    TEXT  geojson file used to clip the input files                                 │
│ --file_type                    TEXT  type of input files: PS | PF                                              │
│ --output_folder                TEXT  output folder, will be created if not present                             │
│ --tmp_folder                   TEXT  scratch folder                                                            │
│ --save_overview_pngs      -sp        Save an overview png for each date.                                       │
│ --save_tiff_results       -st        Save geotif results for each date.                                        │
│ --config_file                  TEXT  Config file, default is dryspy.ini                                      │
│ --startdate                    TEXT  only process file after this date                                         │
│ --enddate                      TEXT  only process file before this date                                        │
│ --help                               Show this message and exit.                                               │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

```

The graph below summarizes the steps taken by the `DrySpy` command.

```mermaid
  flowchart TB;
      r[Stack of \n PF/PS file] --> C[River region];
      A[geojson of region]-->C;
      P[prior water occurrence\n-- result of previous run -- ] --> WDO;
      cf(Configuration</br>File) --> WDO;
      C-->WDO;

      subgraph WDO [Water detection object];
        direction LR;
        D(Normalize</br>nir) --> B(Canny edge</br>detection);  
        B -- Apply</br>buffer -->  H(Determine threshold<br> from buffer using Otsu)
        H --> app(Apply Threshold<br>generate water map);
        app --> DI(Remove small elements<br>Apply erosion and</br>dilation)
        DI --> TEMP[Apply secondary</br>corrections</br>optional]
        TEMP --> W[water</br>map];
        W --> WS[Water</br>statistics];
       
      end
      WDO --> cc[Collect all output\nGenerate extra statistics]
      cc -->  WB(water body map\nother maps)
  
    
      cc --> S[Statistics:\nwet/dry indication etc.]

  
      S --> csv(csv output)
      WB --> T(tiff output)
      WB --> PNG(PNG output)
    
      
```

In the `notebooks/example` folder you can also find the `cli_run_s2.sh` script that runs the command line tool using the Sentinel-2 files in the `notebooks/example/EO_Browser_images/Scott-River` folder.

## Output

The `stats` dictionary (save as a csv after a run) contains the following parameters:

hist_kurtosis
  : kurtosis of the data used in the water detection after scaling

hist_skew
 : skewness of the data used in the water detection after scaling

25
  : 25th percentile

75
  : 75th percentile

diptest
  : results of a diptest of unimodality (Hartigan and Hartigan, 1985) 

hist_slope
  : slope of the histogram at the water/land threshold

diff_in_out
  : difference between the value of the data inside and outside of the main channel (after Wang and Vivoni, 2022)

In addition, the following values for the first 20 (in size) detected water bodies:

wb_area_n
  : area in pixels

wb_mean_n
  : average value of the variable used in water detection in this area

wb_area_length_n
  : length of the longest axis of this area

wb_area_length_short_n
  : length of the shortest axis of this area

wb_area_label_n
  : label of this area

dry_nir_diff
  : dry periods as determined by the nir difference method (True|False)

dry_wb_length
  : dry periods as determined by water body length method (True|False)



## Dry vs running river detection

There are a number of parameters calculated for each timestep (and saves in the `.csv` file) that can be used to determine if the river is flowing or not.

1. The length of individual water bodies detected in the region is calculated and save in the `csv` file as `'wb_area_length_1` to n for each detected water body. We find that the sum of the three largest water bodies is a good measure to use.
2. Based on the work of Wang and Vivoni (2022) an index is calculated based on the value of the band/index used in the water detection inside and outside of the channel. 




## Configuration

``` ini
# Settings for the canny edge detection and otsu thresholding method
[edge_detection]
sigma = 2.5  
# A larger value smooths the image more before finding the edges

# For otsu
classes = 3  
# Number of classes in the otsu thresholding (the lowest class is used)
buffer = 2   
# Buffer to apply after the edge detection to increase the number of sampling points
histogram_scale_maxvalue = 25    
# scaling upper limit for the histogram stretching
histogram_scale_minvalue = 0
threshold_multiplication = 1.0   
# Scale factor to apply to threshold (should not be needed)

# Here we set the thresholds (or percentiles) for nir and ndwi for the water detection
[nir_and_ndwi]
#nir_th = 0.18
#ndwi_th = -0.5847580417346776
nir_th_perc = 5
ndwi_th_perc = 5

# final cleaning of the detected water map (delete small areas)
[data_clean]
min_size = 6 
# Minimum size of the detected elements. Smaller elements are removed
dilation_erosion_size = 6   
# Size of the dilation/erosion kernel used to clean the detected water elements
max_data_value = 0.80
min_data_value = -0.98
min_u_pixel = 0.10
min_u_area = 0.16
min_occur = 0.09
weight_of_prior = 365

[area]
clip_file = clip.geojson   
# should be in the same coordinates as the raster files

[dry_detection]  
# Used in the dryspy cli version only

# Difference in nir inside and outside of the channel
nir_diff_th = -0.05
# Number of water bodies to include in the length calculation
number_wb = 3 
```


## Setting up a new region to monitor

1. Make sure you have access to the PS/PF/S2 data for that region. This will in most cases be a (cloud)storage location holding the data of a subscription
2. Make a geojson (qgis works good for that) of the stretch of river you want to monitor using the same georeferencing as the tiff files with the data. Make sure the area is chosen such that the open water (river) will _not_ cover the majority of the area but is a significant proportion of the area. Try and avoid shadow's and very dark vegetation
3. Copy the example configuration file, give it a new name and edit the values if needed
4. Test running with a single file (timestep) and adjust parameters if required


## References

Donchyts, G., Schellekens, J., Winsemius, H., Eisemann, E., Van de Giesen, N., 2016. A 30 m resolution surface water mask including estimation of positional and thematic differences using landsat 8, srtm and openstreetmap: a case study in the Murray-Darling Basin, Australia. Remote Sensing 8, 386.

J. A. Hartigan. P. M. Hartigan. "The Dip Test of Unimodality." Ann. Statist. 13 (1) 70 - 84, March, 1985. https://doi.org/10.1214/aos/1176346577

Wang, Z., Vivoni, E.R., 2022. Detecting Streamflow in Dryland Rivers Using CubeSats. Geophysical Research Letters 49, e2022GL098729. https://doi.org/10.1029/2022GL098729


CHANGELOG.rst
=========
Changelog
=========

Version 0.1
===========

- initial


tests/conftest.py
"""
    Dummy conftest.py for dryspy.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html
"""

import pytest, os, sys

sys.path.insert(0, os.path.abspath("../"))

from dryspy.co_wdetect import canny_otsu_water_detect


config = {
    "edge_detection": {
        "histogram_scale_maxvalue": 100,
        "histogram_scale_minvalue": 87,
        "sigma": 2.5,
        "classes": 2,
        "threshold_multiplication": 1.0,
        "buffer": 4,
    },
    "data_clean": {
        "min_size": 8,
        "dilation_erosion_size": 4,
        "max_data_value": 0.60,
        "min_data_value": -0.48,
    },
}


def test_init():
    wd = canny_otsu_water_detect(config)

    wd.config["edge_detection"]["histogram_scale_maxvalue"]

    assert wd.config["edge_detection"]["histogram_scale_maxvalue"] == "100"


def test_read_detect():
    import rasterio
    import numpy as np

    wd = canny_otsu_water_detect(config)

    B8_ = rasterio.open(
        "notebooks/example/EO_Browser_images/2019-06-07-00:00_2019-06-07-23:59_Sentinel-2_L1C_B08_(Raw).tiff"
    )

    # Apply the scaling
    B8 = B8_.read() / 100000

    # Waterdetect assumes missing values are set to np.nan so we do that here
    B8 = np.where(B8 <= 0.001, np.nan, B8)

    water = wd.detect_water(
        B8[0, :, :], apply_uncertainty_filter=True, th_direction="larger"
    )

    assert water.sum() == 50206


def test_read_detect_multiple():
    import rasterio
    import numpy as np

    wd = canny_otsu_water_detect(config)

    B8_ = rasterio.open(
        "notebooks/example/EO_Browser_images/2019-06-07-00:00_2019-06-07-23:59_Sentinel-2_L1C_B08_(Raw).tiff"
    )
    B11_ = rasterio.open(
        "notebooks/example/EO_Browser_images/2019-06-07-00:00_2019-06-07-23:59_Sentinel-2_L1C_B11_(Raw).tiff"
    )
    B3_ = rasterio.open(
        "notebooks/example/EO_Browser_images/2019-06-07-00:00_2019-06-07-23:59_Sentinel-2_L1C_B03_(Raw).tiff"
    )
    B4_ = rasterio.open(
        "notebooks/example/EO_Browser_images/2019-06-07-00:00_2019-06-07-23:59_Sentinel-2_L1C_B04_(Raw).tiff"
    )
    # Apply the scaling
    B8 = B8_.read() / 100000
    B11 = B11_.read() / 100000
    B3 = B3_.read() / 100000
    B4 = B4_.read() / 100000

    # Waterdetect assumes missing values are set to np.nan so we do that here
    B8 = np.where(B8 <= 0.001, np.nan, B8)
    B3 = np.where(B3 <= 0.001, np.nan, B3)
    B11 = np.where(B11 <= 0.001, np.nan, B11)
    B4 = np.where(B11 <= 0.001, np.nan, B4)

    # Next we calculate nwdi from swir and nir
    ndwi_swir = (B3 - B11) / (B3 + B11)
    ndvi = (B8 - B4) / (B4 + B8)

    # Waterdetect assumes missing values are set to np.nan so we do that here
    B8 = np.where(B8 <= 0.001, np.nan, B8)

    water = wd.detect_water(
        B8[0, :, :],
        apply_uncertainty_filter=True,
        apply_secondary_filter=True,
        max_data=ndvi[0, :, :],
        min_data=ndwi_swir[0, :, :],
        th_direction="smaller",
    )

    assert water.sum() == 1009313


src/dryspy/__init__.py
"""
***************************************************************************
Copyright Planet Labs PBC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
***************************************************************************
"""

import sys

if sys.version_info[:2] >= (3, 8):
    # TODO: Import directly (no need for conditional) when `python_requires = >= 3.8`
    from importlib.metadata import PackageNotFoundError, version  # pragma: no cover
else:
    from importlib_metadata import PackageNotFoundError, version  # pragma: no cover

try:
    # Change here if project is renamed and does not equal the package name
    dist_name = "dryspy"
    __version__ = version(dist_name)
except PackageNotFoundError:  # pragma: no cover
    __version__ = "unknown"
finally:
    del version, PackageNotFoundError


src/dryspy/co_wdetect.py
"""
***************************************************************************
Copyright Planet Labs PBC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
***************************************************************************
"""

from skimage import morphology
import configparser
from skimage import exposure
import numpy as np
from skimage import feature
from skimage import measure
from skimage.filters import threshold_multiotsu
from skimage.morphology import square
from scipy.stats import kurtosis, skew
import diptest
import pandas as pd


def s_curve(X, a=0.0, b=1.0, c=1.0):
    """s_curve function, transform data into a range

    Input
    -----
        - X input array
        - C determines the steepness or "stepwiseness" of the curve.
          The higher C the sharper the function. A negative C reverses the function.
        - b determines the amplitude of the curve
        - a determines the centre level (default = 0)

    Output
    ------
        - transformed array
    """
    s = 1.0 / (b + np.exp(-c * (X - a)))

    return s


class canny_otsu_water_detect:
    def __init__(self, config, prior_water_occurrence=None):
        """Water detection using canny edge detection and otsu thresholding

        Parameters
        ----------
        config : configparser.ConfigParser or dict
            configparser object with values for the configurable items. If set to None
            default values will be used. If it is specified all the expected sections
            should be present: [edge_detection], [data_clean]. It can be specified as an
            dictionary or a configparser object
        prior_water_occurrence: 2d ndarray floats (optional)
            should have the same dimensions and the input maps used after initialisation
        """

        if config is None:
            self.config = configparser.ConfigParser(allow_no_value=True)
            self.config.read_dict(
                {
                    "edge_detection": {
                        "histogram_scale_maxvalue": 25,
                        "sigma": 2.0,
                        "classes": 3,
                        "threshold_multiplication": 1.0,
                        "buffer": 2,
                    },
                    "data_clean": {"min_size": 6, "dilation_erosion_size": 6},
                }
            )
        else:
            if isinstance(config, dict):
                self.config = configparser.ConfigParser(allow_no_value=True)
                self.config.read_dict(config)
            else:
                self.config = config  # configparser

        self.canny_buffedges = None
        self.canny_edges = None
        self.thresholds = None
        self.water_canny_otsu = None
        self.water_sec = None
        self.water_uncert = None
        self.water_final = None
        self.invalid = None
        self.data = None
        self.peak = None
        """stats 
        Statistics associated with the water detection
        
        """
        self.stats = {}
        self.count = 0
        self.wat_occur = None
        self.avguncertainty = None
        self.prior_wateroccurence = prior_water_occurrence

        self.weight_of_prior = self.config["data_clean"].getint(
            "weight_of_prior", fallback=100
        )  # 100 days
        self.u_pixel = (
            None  # Pixel based relative uncertainty (1 = certain, 0 = very uncertain)
        )
        self.u_pixel_distance = None
        self.u = None  # otsu thresholding estimated algorithm uncertainty
        self.u1 = None  # uncertainty component
        self.u2 = None  # uncertainty component
        self.u3 = None  # uncertainty component

    def _morphology_clean(self, watertoclean: np.array):
        """removes small areas and performs a dilation followed by an erosion
        to clean up the edges

        Parameters
        ----------
        watertoclean : numpy array (2d)
            incoming binary map

        Returns
        -------
        numpy array 2d
            outgoing (cleaned) binary map
        """
        # Remove small objects
        water_cleaned = morphology.remove_small_objects(
            watertoclean.copy(),
            min_size=self.config["data_clean"].getint("min_size", fallback=6),
        )
        # Do dilation and erosion to get rid of small gaps and connect objects that
        # are very close to each other
        de_size = self.config["data_clean"].getint("dilation_erosion_size", fallback=6)
        water_cleaned = morphology.binary_dilation(
            water_cleaned, footprint=np.ones((de_size, de_size))
        )
        water_cleaned = morphology.binary_erosion(
            water_cleaned, footprint=np.ones((de_size, de_size))
        )

        return water_cleaned

    def _canny_edges(self, data: np.array):
        """Apply canny edge detection after rescaling between percentiles
        percentile (default 0 and 25). Next apply a buffer and determine threshold with otsu

        Parameters
        ----------
        data : 2dndarray of floats
            data to use in edge detection
        """

        # Get histogram scaling parameters
        max = self.config["edge_detection"].getint(
            "histogram_scale_maxvalue", fallback=25
        )
        min = self.config["edge_detection"].getint(
            "histogram_scale_minvalue", fallback=0
        )
        # Scale data
        L, H = np.percentile(data[~np.isnan(data)], (min, max))
        data_rescale = exposure.rescale_intensity(data, in_range=(L, H))
        data_rescale[np.isnan(data)] = np.nan

        # Do edge detection on rescaled data
        self.canny_edges = feature.canny(
            data_rescale,
            sigma=self.config["edge_detection"].getfloat("sigma", fallback=2.0),
        )
        # Apply the buffer
        edges_buff = morphology.dilation(
            self.canny_edges,
            footprint=square(
                self.config["edge_detection"].getint("buffer", fallback=2)
            ),
        )
        # Select original data at the edge plus buffer location and determine thresholds
        data_edges = data[edges_buff]

        if data_edges.sum() == 0:
            thresholds = np.array([0.0, 0.0])
        else:
            classes = self.config["edge_detection"].getint("classes", fallback=3)
            thresholds = threshold_multiotsu(data_edges, classes=classes)
            thresholds = np.array(thresholds)

        # TDO: nr op sample points?
        # apply multiplication if configured
        self.canny_buffedges = edges_buff
        self.thresholds = thresholds.copy()
        mult = self.config["edge_detection"].getfloat(
            "threshold_multiplication", fallback=1.0
        )
        self.thresholds = self.thresholds * float(mult)

        return self.thresholds[0]

    def _apply_secundary_filter(
        self, water_in: np.array, max_data: np.array, min_data: np.array
    ):
        """Apply a secundary filter on the detected water
        if these values are not configure in the config the filter wil _NOT_ be applied.
        This can be run _after_ the detect_water method

        [data_clean]
        - max filter using max_data
        - min filter using min_date

        Parameters
        ----------
        water_in : np.array
            initial estimate of water to be filtered
        max_data : np.array
            array with  data to use above which we assume there is no water
            (typically something like an ndvi)
        min_data : np.array
            array with  data to use below which we assume there is no water
            (typically something like an ndwi)
        """
        max_data_value = self.config["data_clean"].getfloat(
            "max_data_value", fallback=-999.0
        )
        min_data_value = self.config["data_clean"].getfloat(
            "min_data_value", fallback=-999.0
        )

        self.water_sec = water_in.copy()
        if max_data_value > -998:
            filt = max_data > max_data_value
            self.water_sec[filt] = 0
        if min_data_value > -998:
            filt = min_data < min_data_value
            self.water_sec[filt] = 0

        if np.sum(self.water_sec) != np.sum(water_in):
            changed = True
        else:
            changed = False

        return self.water_sec.copy(), changed

    def _det_and_apply_uncertainty_filter(self, water_in: np.array):
        """_summary_"""

        if self.count == 1:
            self.sumuncertainty = np.nan_to_num(self.u_pixel.astype(float).copy())
        else:
            self.sumwat = self.sumwat + water_in.astype(int)
            self.sumuncertainty = np.nan_to_num(self.sumuncertainty) + np.nan_to_num(
                self.u_pixel
            )

        self.water_uncert = water_in.copy()

        # First determine average uncertainty per water body.
        min_u_area = self.config["data_clean"].getfloat("min_u_area", fallback=0)
        self.label_areas = measure.label(water_in)
        props = measure.regionprops_table(
            self.label_areas,
            self.u_pixel,
            properties=[
                "label",
                "area",
                "intensity_mean",
                "axis_major_length",
                "axis_minor_length",
            ],
        )
        props = pd.DataFrame(props)
        areas_to_filter = props.where(props["intensity_mean"] < min_u_area).dropna()
        for area in areas_to_filter.iterrows():
            self.water_uncert[area[1]["label"] == self.label_areas] = 0

        # Next filter out points uncertain points based on this timestep
        min_u_pixel = self.config["data_clean"].getfloat("min_u_pixel", fallback=0)

        self.water_uncert[self.u_pixel < min_u_pixel] = 0

        if self.prior_wateroccurence is None:
            self.wat_occur = self.sumwat / self.count
        else:
            self.wat_occur = self.sumwat / (self.count + self.weight_of_prior)

        self.avguncertainty = self.sumuncertainty / self.count

        # Filter out points based on history (can only be done if the object is
        # run multiple times)
        min_occur = self.config["data_clean"].getfloat("min_occur", fallback=0)
        self.water_uncert[self.wat_occur < min_occur] = 0

        if np.sum(self.water_uncert) != np.sum(water_in):
            changed = True
        else:
            changed = False

        return self.water_uncert.copy(), changed

    def _extra_stats(self):
        """Add some extra statistics to the stats dictionary"""
        L, H = np.percentile(
            self.wat_occur[~np.isnan(self.wat_occur)],
            (5, 95),
        )
        self.stats["diff_in_out"] = np.nanmean(
            self.data[self.wat_occur >= H]
        ) - np.nanmean(self.data[self.wat_occur <= L])
        self.label_areas = measure.label(self.water_final)
        props = measure.regionprops_table(
            self.label_areas,
            self.data,
            properties=[
                "label",
                "area",
                "intensity_mean",
                "axis_major_length",
                "axis_minor_length",
            ],
        )
        props = pd.DataFrame(props)
        props.sort_values(by=["area"], ascending=False, inplace=True)
        i = 0
        for prop in props.iterrows():
            i = i + 1
            if i <= 20:
                self.stats["wb_area_" + str(i)] = prop[1].area
                self.stats["wb_mean_" + str(i)] = prop[1].intensity_mean
                self.stats["wb_area_length_" + str(i)] = prop[1].axis_major_length
                self.stats["wb_area_length_short_" + str(i)] = prop[1].axis_minor_length
                self.stats["wb_area_label_" + str(i)] = prop[1].label

    def detect_water(
        self,
        data: np.array,
        th_direction="smaller",
        apply_secondary_filter=False,
        max_data: np.array = None,
        min_data: np.array = None,
        apply_uncertainty_filter=False,
    ):
        """Calls the canny/otsu water detection method using the 2darray in data,
        sets the output in object-wide variables, estimates relative uncertainty
        and does some post-processing.

        Parameters
        ----------
        data : 2d array of floats non valid data should be indicated as np.nan
            grid to use in water detection
        th_direction: string
            direction of threshold smaller|larger (e.g. for ndwi you would chose larger
            (water is in the high range), for nir you would chose smaller

        Returns
        -------
        2d boolean array
            water grid (1 === water, 0 == no water)
        """
        self.data = data.copy()
        if data.dtype != float:
            raise TypeError(f"data should be of type float, got: '{data.dtype}'")

        # step 1 edge detection
        self._canny_edges(data)

        # First calculate some stats for the input data
        buffdata = self.data[self.canny_buffedges].copy()
        self.stats["hist_kurtosis"] = kurtosis(buffdata[~np.isnan(buffdata)].flatten())
        self.stats["hist_skew"] = skew(buffdata[~np.isnan(buffdata)].flatten())
        if len(~np.isnan(buffdata).flatten()) > 0:
            self.stats["25"], self.stats["75"] = np.nanpercentile(
                buffdata, [25.0, 75.0]
            )
        else:
            self.stats["25"] = np.nan
            self.stats["75"] = np.nan

        if np.sum((buffdata[~np.isnan(buffdata)].flatten())) > 0:
            self.stats["diptest"] = diptest.dipstat(
                buffdata[~np.isnan(buffdata)].flatten()
            )
        else:
            self.stats["diptest"] = np.nan
        # Convert kurt and skew to an apparent uncertainty in the otsu thresholding
        self.u1 = s_curve(self.stats["hist_skew"], a=2, b=1, c=-1)
        self.u2 = s_curve(self.stats["hist_kurtosis"], a=2, b=1, c=-1)

        # the slope if the histogram at the threshold is another source of
        # uncertainty. Here we determine the slope and use the average of three points
        # TODO: also add for original hist (non sampled)
        hist = np.histogram(buffdata, bins="auto", density=True)
        difs = np.absolute(self.thresholds[0] - hist[1])
        idx = difs.argmin()  # index closest to threshold
        slopes = np.diff(hist[0])
        if np.isnan(hist[0][[0]]):
            self.slope_at_threshold = np.nan
        else:
            self.slope_at_threshold = np.nanmean(
                np.abs(slopes[idx - 1 : idx + 1]) / np.nanmean(np.diff(hist[1]))
            )
        self.stats["hist_slope"] = self.slope_at_threshold
        self.u3 = s_curve(self.slope_at_threshold, a=35000, b=1, c=-0.00007)

        # We take the mean of the three relative uncertainties as they are
        # different estimates of the same uncertainty
        self.u = (self.u1 + self.u2 + self.u3) / 3.0

        if th_direction == "larger":
            self.water_canny_otsu = data > self.thresholds[0]
        else:
            self.water_canny_otsu = data < self.thresholds[0]

        self.water_canny_otsu = self._morphology_clean(self.water_canny_otsu)

        self.water_final = self.water_canny_otsu.copy()
        # estimate pixel bases uncertainty using the distance to the threshold,
        # max out at 1
        self.u_pixel_distance = np.minimum(
            1.0,
            np.absolute(self.data - self.thresholds[0])
            / (self.stats["75"] - self.stats["25"]),
        )
        self.u_pixel = self.u_pixel_distance * self.u
        self.u_pixel[self.water_final == 0] = np.nan

        self.sec_changed = False
        self.uncert_changed = False
        if apply_secondary_filter:
            self.water_final, self.sec_changed = self._apply_secundary_filter(
                self.water_final, max_data, min_data
            )

        self.count = self.count + 1  # for temporal occurrence
        if self.count == 1:
            if self.prior_wateroccurence is None:
                self.sumwat = self.water_final.copy()
                self.wat_occur = self.sumwat.copy().astype(int)
            else:
                self.sumwat = self.prior_wateroccurence * self.weight_of_prior
                self.wat_occur = self.prior_wateroccurence

        if apply_uncertainty_filter:
            (
                self.water_final,
                self.uncert_changed,
            ) = self._det_and_apply_uncertainty_filter(self.water_final)

        self.water_final = self._morphology_clean(self.water_final)

        # fill the stats dictionary with extra statistics
        self._extra_stats()

        return self.water_final
